src/main/java/org/opensearch/geospatial/ip2geo/common/DatasourceFacade.java

```diff
@@ -120,4 +127,71 @@ public void onFailure(final Exception e) {
             }
         });
     }
+
+    /**
+     * Get datasources from an index {@code DatasourceExtension.JOB_INDEX_NAME}
+     * @param names the array of datasource names
+     * @param actionListener the action listener
+     */
+    public void getDatasources(final String[] names, final ActionListener<List<Datasource>> actionListener) {
+        client.prepareMultiGet().add(DatasourceExtension.JOB_INDEX_NAME, names).execute(new ActionListener<>() {
+            @Override
+            public void onResponse(final MultiGetResponse response) {
+                List<Datasource> datasources = new ArrayList<>(response.getResponses().length);
+                for (MultiGetItemResponse item : response.getResponses()) {
+                    if (item.getResponse() != null && item.getResponse().isExists()) {
+                        try {
+                            XContentParser parser = XContentHelper.createParser(
+                                NamedXContentRegistry.EMPTY,
+                                LoggingDeprecationHandler.INSTANCE,
+                                item.getResponse().getSourceAsBytesRef()
+                            );
+                            datasources.add(Datasource.PARSER.parse(parser, null));
+                        } catch (IOException e) {
+                            actionListener.onFailure(e);
+                            return;
+                        }
+                    }
+                }
+                actionListener.onResponse(datasources);
+            }
+
+            @Override
+            public void onFailure(final Exception e) {
+                actionListener.onFailure(e);
+            }
+        });
+    }
+
+    /**
+     * Get all datasources from an index {@code DatasourceExtension.JOB_INDEX_NAME}
+     * @param actionListener the action listener
+     */
+    public void getAllDatasources(final ActionListener<List<Datasource>> actionListener) {
+        client.prepareSearch(DatasourceExtension.JOB_INDEX_NAME).setQuery(QueryBuilders.matchAllQuery()).execute(new ActionListener<>() {
+            @Override
+            public void onResponse(final SearchResponse searchResponse) {
+                List<Datasource> datasources = new ArrayList<>(searchResponse.getHits().getHits().length);
+                for (SearchHit searchHit : searchResponse.getHits().getHits()) {
+                    try {
+                        XContentParser parser = XContentHelper.createParser(
```

Comment:
I can see some common code across these functions. See if we can abstract them out in  a single function.